File: crates/kuben-core/src/ops/outcome.rs
//! What a build Job's observed state means for its attempt (ADR-028).
//!
//! The executor reads the Job and its pod and hands the facts to
//! [`classify`], a pure function, so every failure class (out of memory, a
//! full disk, the deadline, a lost node, a bad source, bad output) is decided
//! and tested in one place. A pod's termination message is only a hint: a
//! reported digest is checked against the registry before the attempt
//! succeeds.

use serde::{Deserialize, Serialize};

use crate::artifact::Digest;

/// Name of the init container that fetches the source.
pub const FETCH_CONTAINER: &str = "fetch";
/// Name of the container that builds and pushes.
pub const BUILD_CONTAINER: &str = "build";
/// Name of the container that scans the pushed image (M4.6). It runs after
/// the build and never decides the build's outcome: a scan that fails is
/// recorded as unavailable.
pub const SCAN_CONTAINER: &str = "scan";
/// How long a build pod may stay unschedulable before the attempt fails.
pub const UNSCHEDULABLE_LIMIT_SECS: u64 = 600;
const DETAIL_MAX: usize = 1024;

/// Why an attempt failed. The code is stored and shown; only a lost worker is
/// retried automatically, as a new attempt.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum BuildFailure {
    /// The build itself failed (a Dockerfile step, a compiler error).
    BuildError,
    /// The build exceeded its memory limit.
    OutOfMemory,
    /// The build exceeded its ephemeral-storage limit or the disk filled up.
    DiskFull,
    /// The build ran past its deadline.
    DeadlineExceeded,
    /// The node or pod disappeared under the build.
    LostWorker,
    /// The repository or commit could not be fetched.
    SourceUnavailable,
    /// The build finished without a valid, verifiable output.
    OutputRejected,
    /// No node could take the build within the limit.
    Unschedulable,
    /// The provider refused the credentials or the installation is suspended.
    CredentialsRefused,
    /// The Job could not be created with its budget.
    InvalidBudget,
}

impl BuildFailure {
    #[must_use]
    pub const fn code(self) -> &'static str {
        match self {
            Self::BuildError => "BuildError",
            Self::OutOfMemory => "OutOfMemory",
            Self::DiskFull => "DiskFull",
            Self::DeadlineExceeded => "DeadlineExceeded",
            Self::LostWorker => "LostWorker",
            Self::SourceUnavailable => "SourceUnavailable",
            Self::OutputRejected => "OutputRejected",
            Self::Unschedulable => "Unschedulable",
            Self::CredentialsRefused => "CredentialsRefused",
            Self::InvalidBudget => "InvalidBudget",
        }
    }

    /// Infrastructure failures a new attempt may not repeat.
    #[must_use]
    pub const fn retryable(self) -> bool {
        matches!(self, Self::LostWorker)
    }

    /// A short explanation with the next step for the user.
    #[must_use]
    pub const fn explain(self) -> &'static str {
        match self {
            Self::BuildError => "the build failed; see the build log",
            Self::OutOfMemory => {
                "the build ran out of memory; raise the build memory limit or use a build node"
            }
            Self::DiskFull => {
                "the build ran out of disk; raise the build storage limit or shrink the context"
            }
            Self::DeadlineExceeded => {
                "the build ran past its deadline; raise the deadline or speed the build up"
            }
            Self::LostWorker => "the build pod or its node was lost; the build is retried",
            Self::SourceUnavailable => {
                "the commit could not be fetched; check the repository and the app's access"
            }
            Self::OutputRejected => "the build reported no verifiable image",
            Self::Unschedulable => "no node had room for the build; add a build node or lower its requests",
            Self::CredentialsRefused => "the Git provider refused access; check the installation",
            Self::InvalidBudget => "the build budget is invalid",
        }
    }
}

/// A terminated container as the pod status reports it.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct ContainerExit {
    pub name: String,
    pub exit_code: i32,
    pub reason: Option<String>,
    pub message: Option<String>,
}

/// The facts about the build pod.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct PodObservation {
    /// `Pending`, `Running`, `Succeeded`, `Failed` or `Unknown`.
    pub phase: String,
    pub reason: Option<String>,
    pub message: Option<String>,
    /// Containers (init or not) that have terminated.
    pub exits: Vec<ContainerExit>,
    /// Containers currently running.
    pub running: Vec<String>,
    /// How long the pod has been unschedulable, if it is.
    pub unschedulable_secs: Option<u64>,
}

/// The facts about the build Job.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct JobObservation {
    pub exists: bool,
    pub succeeded: bool,
    pub failed: bool,
    /// Reason of the Job's `Failed` condition (`DeadlineExceeded`,
    /// `BackoffLimitExceeded`).
    pub failed_reason: Option<String>,
    pub pod: Option<PodObservation>,
}

/// What the build container writes to its termination log.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct BuildReport {
    pub digest: Digest,
    /// The strategy the pod chose (`dockerfile` or `railpack`).
    #[serde(default)]
    pub strategy: String,
}

/// Where a build stands, from its Job.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum JobVerdict {
    /// Created, not started yet.
    Pending,
    /// Fetching the source.
    Fetching,
    /// Building (and pushing, which is the same command).
    Building,
    /// Finished with a report still to verify.
    Finished(BuildReport),
    Failed {
        failure: BuildFailure,
        detail: String,
    },
    /// The Job does not exist.
    Gone,
}

fn bounded(text: &str) -> String {
    let text = text.trim();
    if text.len() <= DETAIL_MAX {
        return text.to_owned();
    }
    let mut end = DETAIL_MAX;
    while !text.is_char_boundary(end) {
        end -= 1;
    }
    text[..end].to_owned()
}

fn failed(failure: BuildFailure, detail: impl AsRef<str>) -> JobVerdict {
    let detail = detail.as_ref();
    JobVerdict::Failed {
        failure,
        detail: if detail.trim().is_empty() {
            failure.explain().to_owned()
        } else {
            bounded(detail)
        },
    }
}

fn mentions_disk(text: Option<&str>) -> bool {
    text.is_some_and(|t| {
        let t = t.to_ascii_lowercase();
        t.contains("no space left on device")
            || t.contains("ephemeral-storage")
            || t.contains("ephemeral local storage")
    })
}

/// The meaning of `job` for the attempt, by the precedence documented on
/// each rule: resource kills first, then the deadline, the source, the
/// build, the output and the worker.
#[must_use]
pub fn classify(job: &JobObservation) -> JobVerdict {
    if !job.exists {
        return JobVerdict::Gone;
    }
    let pod = job.pod.as_ref();
    let all_exits = pod.map_or(&[][..], |p| p.exits.as_slice());
    let scan_exit = all_exits.iter().find(|e| e.name == SCAN_CONTAINER);
    let exits: Vec<&ContainerExit> = all_exits.iter().filter(|e| e.name != SCAN_CONTAINER).collect();
    let failed_exit = exits.iter().copied().find(|e| e.exit_code != 0);
    let report = || {
        exits
            .iter()
            .find(|e| e.name == BUILD_CONTAINER && e.exit_code == 0)
            .and_then(|e| e.message.as_deref())
            .and_then(|m| serde_json::from_str::<BuildReport>(m.trim()).ok())
    };

    // 1. The kernel's OOM killer is unambiguous.
    if exits.iter().any(|e| e.reason.as_deref() == Some("OOMKilled")) {
        return failed(BuildFailure::OutOfMemory, "");
    }
    // 2. An eviction for ephemeral storage, or a build that saw ENOSPC.
    let evicted = pod.is_some_and(|p| p.reason.as_deref() == Some("Evicted"));
    let disk = (evicted && mentions_disk(pod.and_then(|p| p.message.as_deref())))
        || exits
            .iter()
            .any(|e| e.exit_code != 0 && mentions_disk(e.message.as_deref()));
    if disk {
        return failed(
            BuildFailure::DiskFull,
            pod.and_then(|p| p.message.clone()).unwrap_or_default(),
        );
    }
    // 3. The Job's activeDeadlineSeconds.
    let deadline = job.failed_reason.as_deref() == Some("DeadlineExceeded")
        || pod.is_some_and(|p| p.reason.as_deref() == Some("DeadlineExceeded"));
    if deadline {
        return failed(BuildFailure::DeadlineExceeded, "");
    }
    // 4. Any other eviction or a vanished node is the worker, not the build.
    let lost =
        evicted || pod.is_some_and(|p| p.phase == "Unknown" || p.reason.as_deref() == Some("NodeLost"));
    if lost {
        return failed(
            BuildFailure::LostWorker,
            pod.and_then(|p| p.message.clone()).unwrap_or_default(),
        );
    }
    // 5. The fetch init container failed.
    if let Some(exit) = failed_exit.filter(|e| e.name == FETCH_CONTAINER) {
        return failed(
            BuildFailure::SourceUnavailable,
            exit.message.as_deref().unwrap_or_default(),
        );
    }
    // 6. The build container failed.
    if let Some(exit) = failed_exit {
        let detail = exit
            .message
            .clone()
            .unwrap_or_else(|| format!("{} exited with code {}", exit.name, exit.exit_code));
        return failed(BuildFailure::BuildError, detail);
    }
    // 7. Success needs a parsable report from the build container; a scan
    // that ended, however it ended, does not change that.
    if job.succeeded || (job.failed && scan_exit.is_some()) {
        return match report() {
            Some(report) => JobVerdict::Finished(report),
            None => failed(
                BuildFailure::OutputRejected,
                "the build container reported no image digest",
            ),
        };
    }
    // 8. A failed Job without an explanation lost its pod.
    if job.failed {
        return failed(
            BuildFailure::LostWorker,
            job.failed_reason
                .as_deref()
                .unwrap_or("the build pod disappeared"),
        );
    }
    let Some(pod) = pod else {
        return JobVerdict::Pending;
    };
    if pod
        .unschedulable_secs
        .is_some_and(|s| s >= UNSCHEDULABLE_LIMIT_SECS)
    {
        return failed(
            BuildFailure::Unschedulable,
            pod.message.as_deref().unwrap_or_default(),
        );
    }
    if pod
        .running
        .iter()
        .any(|c| c == BUILD_CONTAINER || c == SCAN_CONTAINER)
    {
        JobVerdict::Building
    } else if pod.running.iter().any(|c| c == FETCH_CONTAINER)
        || exits.iter().any(|e| e.name == FETCH_CONTAINER)
    {
        JobVerdict::Fetching
    } else {
        JobVerdict::Pending
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const DIGEST: &str = "sha256:0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";

    fn exit(name: &str, code: i32, reason: Option<&str>, message: Option<&str>) -> ContainerExit {
        ContainerExit {
            name: name.into(),
            exit_code: code,
            reason: reason.map(Into::into),
            message: message.map(Into::into),
        }
    }

    fn job(pod: PodObservation) -> JobObservation {
        JobObservation {
            exists: true,
            pod: Some(pod),
            ..JobObservation::default()
        }
    }

    fn failure(verdict: &JobVerdict) -> Option<BuildFailure> {
        match verdict {
            JobVerdict::Failed { failure, .. } => Some(*failure),
            _ => None,
        }
    }

    #[test]
    fn a_missing_job_is_gone() {
        assert_eq!(classify(&JobObservation::default()), JobVerdict::Gone);
    }

    #[test]
    fn progress_follows_the_running_container() {
        let pending = job(PodObservation {
            phase: "Pending".into(),
            ..PodObservation::default()
        });
        assert_eq!(classify(&pending), JobVerdict::Pending);
        let fetching = job(PodObservation {
            phase: "Pending".into(),
            running: vec![FETCH_CONTAINER.into()],
            ..PodObservation::default()
        });
        assert_eq!(classify(&fetching), JobVerdict::Fetching);
        let building = job(PodObservation {
            phase: "Running".into(),
            exits: vec![exit(FETCH_CONTAINER, 0, Some("Completed"), None)],
            running: vec![BUILD_CONTAINER.into()],
            ..PodObservation::default()
        });
        assert_eq!(classify(&building), JobVerdict::Building);
        let no_pod_yet = JobObservation {
            exists: true,
            ..JobObservation::default()
        };
        assert_eq!(classify(&no_pod_yet), JobVerdict::Pending);
    }

    #[test]
    fn success_needs_a_valid_report() {
        let mut done = job(PodObservation {
            phase: "Succeeded".into(),
            exits: vec![
                exit(FETCH_CONTAINER, 0, None, None),
                exit(
                    BUILD_CONTAINER,
                    0,
                    Some("Completed"),
                    Some(&format!(r#"{{"digest":"{DIGEST}","strategy":"railpack"}}"#)),
                ),
            ],
            ..PodObservation::default()
        });
        done.succeeded = true;
        let JobVerdict::Finished(report) = classify(&done) else {
            panic!("not finished");
        };
        assert_eq!(report.digest.as_str(), DIGEST);
        assert_eq!(report.strategy, "railpack");

        let mut forged = done.clone();
        forged.pod.as_mut().expect("pod").exits[1].message = Some(r#"{"digest":"sha256:nope"}"#.into());
        assert_eq!(failure(&classify(&forged)), Some(BuildFailure::OutputRejected));
        let mut silent = done;
        silent.pod.as_mut().expect("pod").exits[1].message = None;
        assert_eq!(failure(&classify(&silent)), Some(BuildFailure::OutputRejected));
    }

    #[test]
    fn out_of_memory_wins_over_a_failed_job() {
        let mut oom = job(PodObservation {
            phase: "Failed".into(),
            exits: vec![exit(BUILD_CONTAINER, 137, Some("OOMKilled"), None)],
            ..PodObservation::default()
        });
        oom.failed = true;
        oom.failed_reason = Some("BackoffLimitExceeded".into());
        let verdict = classify(&oom);
        assert_eq!(failure(&verdict), Some(BuildFailure::OutOfMemory));
        assert!(matches!(verdict, JobVerdict::Failed { detail, .. } if detail.contains("memory")));
    }

    #[test]
    fn a_full_disk_is_told_apart_from_other_evictions() {
        let disk = job(PodObservation {
            phase: "Failed".into(),
            reason: Some("Evicted".into()),
            message: Some(
                "Pod ephemeral local storage usage exceeds the total limit of containers 2Gi.".into(),
            ),
            ..PodObservation::default()
        });
        assert_eq!(failure(&classify(&disk)), Some(BuildFailure::DiskFull));
        let enospc = job(PodObservation {
            phase: "Failed".into(),
            exits: vec![exit(
                BUILD_CONTAINER,
                1,
                Some("Error"),
                Some("write /tmp/x: no space left on device"),
            )],
            ..PodObservation::default()
        });
        assert_eq!(failure(&classify(&enospc)), Some(BuildFailure::DiskFull));
        let pressure = job(PodObservation {
            phase: "Failed".into(),
            reason: Some("Evicted".into()),
            message: Some("The node was low on resource: memory.".into()),
            ..PodObservation::default()
        });
        assert_eq!(failure(&classify(&pressure)), Some(BuildFailure::LostWorker));
        assert!(BuildFailure::LostWorker.retryable());
        assert!(!BuildFailure::DiskFull.retryable());
    }

    #[test]
    fn the_deadline_is_reported_as_such() {
        let late = JobObservation {
            exists: true,
            failed: true,
            failed_reason: Some("DeadlineExceeded".into()),
            pod: None,
            ..JobObservation::default()
        };
        assert_eq!(failure(&classify(&late)), Some(BuildFailure::DeadlineExceeded));
    }

    #[test]
    fn fetch_and_build_errors_keep_their_message() {
        let fetch = job(PodObservation {
            phase: "Failed".into(),
            exits: vec![exit(
                FETCH_CONTAINER,
                128,
                Some("Error"),
                Some("fatal: reference is not a tree"),
            )],
            ..PodObservation::default()
        });
        assert_eq!(
            classify(&fetch),
            JobVerdict::Failed {
                failure: BuildFailure::SourceUnavailable,
                detail: "fatal: reference is not a tree".into()
            }
        );
        let build = job(PodObservation {
            phase: "Failed".into(),
            exits: vec![
                exit(FETCH_CONTAINER, 0, None, None),
                exit(BUILD_CONTAINER, 1, Some("Error"), Some(&"e".repeat(5000))),
            ],
            ..PodObservation::default()
        });
        let verdict = classify(&build);
        assert!(
            matches!(&verdict, JobVerdict::Failed { failure: BuildFailure::BuildError, detail } if detail.len() == 1024)
        );
    }

    #[test]
    fn the_scan_never_decides_the_build() {
        let report = format!(r#"{{"digest":"{DIGEST}","strategy":"dockerfile"}}"#);
        let scanning = job(PodObservation {
            phase: "Running".into(),
            exits: vec![
                exit(FETCH_CONTAINER, 0, None, None),
                exit(BUILD_CONTAINER, 0, Some("Completed"), Some(&report)),
            ],
            running: vec![SCAN_CONTAINER.into()],
            ..PodObservation::default()
        });
        assert_eq!(classify(&scanning), JobVerdict::Building);
        for scan in [
            exit(SCAN_CONTAINER, 0, Some("Completed"), Some(r#"{"status":"ok"}"#)),
            exit(SCAN_CONTAINER, 137, Some("OOMKilled"), None),
            exit(SCAN_CONTAINER, 1, Some("Error"), Some("no space left on device")),
        ] {
            let failed = scan.exit_code != 0;
            let done = JobObservation {
                exists: true,
                succeeded: !failed,
                failed,
                failed_reason: failed.then(|| "BackoffLimitExceeded".to_owned()),
                pod: Some(PodObservation {
                    phase: if failed { "Failed" } else { "Succeeded" }.into(),
                    exits: vec![
                        exit(FETCH_CONTAINER, 0, None, None),
                        exit(BUILD_CONTAINER, 0, Some("Completed"), Some(&report)),
                        scan.clone(),
                    ],
                    ..PodObservation::default()
                }),
            };
            assert!(
                matches!(classify(&done), JobVerdict::Finished(r) if r.digest.as_str() == DIGEST),
                "{scan:?}"
            );
        }
    }

    #[test]
    fn lost_pods_and_unschedulable_builds() {
        let vanished = JobObservation {
            exists: true,
            failed: true,
            ..JobObservation::default()
        };
        assert_eq!(failure(&classify(&vanished)), Some(BuildFailure::LostWorker));
        let unknown = job(PodObservation {
            phase: "Unknown".into(),
            ..PodObservation::default()
        });
        assert_eq!(failure(&classify(&unknown)), Some(BuildFailure::LostWorker));
        let waiting = job(PodObservation {
            phase: "Pending".into(),
            unschedulable_secs: Some(UNSCHEDULABLE_LIMIT_SECS - 1),
            ..PodObservation::default()
        });
        assert_eq!(classify(&waiting), JobVerdict::Pending);
        let stuck = job(PodObservation {
            phase: "Pending".into(),
            unschedulable_secs: Some(UNSCHEDULABLE_LIMIT_SECS),
            message: Some("0/1 nodes are available: 1 Insufficient memory.".into()),
            ..PodObservation::default()
        });
        assert_eq!(failure(&classify(&stuck)), Some(BuildFailure::Unschedulable));
    }
}

File: go/hub/internal/core/ops/outcome/outcome.go
// Package outcome says what a build Job's observed state means for its
// attempt (ADR-028). It replaces the Rust module
// kuben-core/src/ops/outcome.rs.
//
// The executor reads the Job and its pod and hands the facts to [Classify],
// a pure function, so every failure class (out of memory, a full disk, the
// deadline, a lost node, a bad source, bad output) is decided and tested in
// one place. A pod's termination message is only a hint: a reported digest
// is checked against the registry before the attempt succeeds.
package outcome

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"unicode/utf8"

	"github.com/Teamtem-dev/kuben/go/hub/internal/core/artifact"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/kerr"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/opt"
)

const (
	// FetchContainer is the name of the init container that fetches the source.
	FetchContainer = "fetch"
	// BuildContainer is the name of the container that builds and pushes.
	BuildContainer = "build"
	// ScanContainer is the name of the container that scans the pushed image
	// (M4.6). It runs after the build and never decides the build's outcome: a
	// scan that fails is recorded as unavailable.
	ScanContainer = "scan"
	// UnschedulableLimitSecs is how long a build pod may stay unschedulable
	// before the attempt fails.
	UnschedulableLimitSecs uint64 = 600

	detailMax = 1024
)

// BuildFailure is why an attempt failed. The code is stored and shown; only a
// lost worker is retried automatically, as a new attempt.
type BuildFailure string

// The failures.
const (
	// BuildError: the build itself failed (a Dockerfile step, a compiler error).
	BuildError BuildFailure = "BuildError"
	// OutOfMemory: the build exceeded its memory limit.
	OutOfMemory BuildFailure = "OutOfMemory"
	// DiskFull: the build exceeded its ephemeral-storage limit or the disk
	// filled up.
	DiskFull BuildFailure = "DiskFull"
	// DeadlineExceeded: the build ran past its deadline.
	DeadlineExceeded BuildFailure = "DeadlineExceeded"
	// LostWorker: the node or pod disappeared under the build.
	LostWorker BuildFailure = "LostWorker"
	// SourceUnavailable: the repository or commit could not be fetched.
	SourceUnavailable BuildFailure = "SourceUnavailable"
	// OutputRejected: the build finished without a valid, verifiable output.
	OutputRejected BuildFailure = "OutputRejected"
	// Unschedulable: no node could take the build within the limit.
	Unschedulable BuildFailure = "Unschedulable"
	// CredentialsRefused: the provider refused the credentials or the
	// installation is suspended.
	CredentialsRefused BuildFailure = "CredentialsRefused"
	// InvalidBudget: the Job could not be created with its budget.
	InvalidBudget BuildFailure = "InvalidBudget"
)

// ParseBuildFailure reads a failure code as stored.
func ParseBuildFailure(s string) (BuildFailure, error) {
	switch f := BuildFailure(s); f {
	case BuildError, OutOfMemory, DiskFull, DeadlineExceeded, LostWorker, SourceUnavailable, OutputRejected,
		Unschedulable, CredentialsRefused, InvalidBudget:
		return f, nil
	}
	return "", kerr.New(kerr.Validation, "unknown build failure `%s`", s)
}

// String is the stored code.
func (f BuildFailure) String() string { return string(f) }

// UnmarshalText refuses a failure that does not exist.
func (f *BuildFailure) UnmarshalText(text []byte) error {
	parsed, err := ParseBuildFailure(string(text))
	if err != nil {
		return err
	}
	*f = parsed
	return nil
}

// Retryable reports whether this is an infrastructure failure a new attempt
// may not repeat.
func (f BuildFailure) Retryable() bool { return f == LostWorker }

// Explain is a short explanation with the next step for the user; empty for
// an unknown failure.
func (f BuildFailure) Explain() string {
	switch f {
	case BuildError:
		return "the build failed; see the build log"
	case OutOfMemory:
		return "the build ran out of memory; raise the build memory limit or use a build node"
	case DiskFull:
		return "the build ran out of disk; raise the build storage limit or shrink the context"
	case DeadlineExceeded:
		return "the build ran past its deadline; raise the deadline or speed the build up"
	case LostWorker:
		return "the build pod or its node was lost; the build is retried"
	case SourceUnavailable:
		return "the commit could not be fetched; check the repository and the app's access"
	case OutputRejected:
		return "the build reported no verifiable image"
	case Unschedulable:
		return "no node had room for the build; add a build node or lower its requests"
	case CredentialsRefused:
		return "the Git provider refused access; check the installation"
	case InvalidBudget:
		return "the build budget is invalid"
	}
	return ""
}

// ContainerExit is a terminated container as the pod status reports it.
type ContainerExit struct {
	Name     string
	ExitCode int32
	Reason   opt.Val[string]
	Message  opt.Val[string]
}

// PodObservation has the facts about the build pod.
type PodObservation struct {
	// Phase is `Pending`, `Running`, `Succeeded`, `Failed` or `Unknown`.
	Phase   string
	Reason  opt.Val[string]
	Message opt.Val[string]
	// Exits are the containers (init or not) that have terminated.
	Exits []ContainerExit
	// Running are the containers currently running.
	Running []string
	// UnschedulableSecs is how long the pod has been unschedulable, if it is.
	UnschedulableSecs opt.Val[uint64]
}

// JobObservation has the facts about the build Job.
type JobObservation struct {
	Exists    bool
	Succeeded bool
	Failed    bool
	// FailedReason is the reason of the Job's `Failed` condition
	// (`DeadlineExceeded`, `BackoffLimitExceeded`).
	FailedReason opt.Val[string]
	Pod          opt.Val[PodObservation]
}

// BuildReport is what the build container writes to its termination log.
type BuildReport struct {
	Digest artifact.Digest `json:"digest"`
	// Strategy is the strategy the pod chose (`dockerfile` or `railpack`);
	// empty when the report does not say.
	Strategy string `json:"strategy"`
}

// UnmarshalJSON requires a valid digest, which a plain decode would leave
// zero when the field is missing or null.
func (r *BuildReport) UnmarshalJSON(data []byte) error {
	var raw struct {
		Digest   opt.Val[artifact.Digest] `json:"digest"`
		Strategy json.RawMessage          `json:"strategy"`
	}
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	digest, ok := raw.Digest.Get()
	if !ok {
		return errors.New("build report: missing field `digest`")
	}
	strategy := ""
	if raw.Strategy != nil {
		if err := json.Unmarshal(raw.Strategy, &strategy); err != nil || string(raw.Strategy) == "null" {
			return errors.New("build report: `strategy` is not a string")
		}
	}
	*r = BuildReport{Digest: digest, Strategy: strategy}
	return nil
}

// JobVerdict is where a build stands, from its Job.
//
//sumtype:decl
type JobVerdict interface{ jobVerdict() }

type (
	// Pending: created, not started yet.
	Pending struct{}
	// Fetching the source.
	Fetching struct{}
	// Building (and pushing, which is the same command).
	Building struct{}
	// Finished with a report still to verify.
	Finished struct{ Report BuildReport }
	// Failed for a reason, with the detail to show.
	Failed struct {
		Failure BuildFailure
		Detail  string
	}
	// Gone: the Job does not exist.
	Gone struct{}
)

func (Pending) jobVerdict()  {}
func (Fetching) jobVerdict() {}
func (Building) jobVerdict() {}
func (Finished) jobVerdict() {}
func (Failed) jobVerdict()   {}
func (Gone) jobVerdict()     {}

// bounded trims text and cuts it to detailMax bytes without splitting a rune.
func bounded(text string) string {
	text = strings.TrimSpace(text)
	if len(text) <= detailMax {
		return text
	}
	end := detailMax
	for end > 0 && !utf8.RuneStart(text[end]) {
		end--
	}
	return text[:end]
}

func failed(failure BuildFailure, detail string) JobVerdict {
	if strings.TrimSpace(detail) == "" {
		return Failed{Failure: failure, Detail: failure.Explain()}
	}
	return Failed{Failure: failure, Detail: bounded(detail)}
}

func mentionsDisk(text opt.Val[string]) bool {
	t, ok := text.Get()
	if !ok {
		return false
	}
	t = asciiLower(t)
	return strings.Contains(t, "no space left on device") ||
		strings.Contains(t, "ephemeral-storage") ||
		strings.Contains(t, "ephemeral local storage")
}

// asciiLower lowers A-Z only, so no other letter can come to match.
func asciiLower(s string) string {
	return strings.Map(func(r rune) rune {
		if r >= 'A' && r <= 'Z' {
			return r + ('a' - 'A')
		}
		return r
	}, s)
}

func is(o opt.Val[string], s string) bool {
	v, ok := o.Get()
	return ok && v == s
}

// Classify is the meaning of job for the attempt, by the precedence
// documented on each rule: resource kills first, then the deadline, the
// source, the build, the output and the worker.
func Classify(job JobObservation) JobVerdict {
	if !job.Exists {
		return Gone{}
	}
	pod, hasPod := job.Pod.Get()
	exits := make([]ContainerExit, 0, len(pod.Exits))
	scanEnded := false
	for _, e := range pod.Exits {
		if e.Name == ScanContainer {
			scanEnded = true
			continue
		}
		exits = append(exits, e)
	}
	if verdict, ok := classifyFailure(job, pod, exits); ok {
		return verdict
	}
	// 7. Success needs a parsable report from the build container; a scan
	// that ended, however it ended, does not change that.
	if job.Succeeded || (job.Failed && scanEnded) {
		if report, ok := reportOf(exits); ok {
			return Finished{Report: report}
		}
		return failed(OutputRejected, "the build container reported no image digest")
	}
	// 8. A failed Job without an explanation lost its pod.
	if job.Failed {
		return failed(LostWorker, job.FailedReason.Or("the build pod disappeared"))
	}
	if !hasPod {
		return Pending{}
	}
	return progress(pod, exits)
}

// classifyFailure applies rules 1 to 6 to a Job that exists. The zero pod
// stands for no pod: none of its facts is set. exits leaves the scan out.
func classifyFailure(job JobObservation, pod PodObservation, exits []ContainerExit) (JobVerdict, bool) {
	if verdict, ok := resourceKill(pod, exits); ok {
		return verdict, true
	}
	// 3. The Job's activeDeadlineSeconds.
	if is(job.FailedReason, "DeadlineExceeded") || is(pod.Reason, "DeadlineExceeded") {
		return failed(DeadlineExceeded, ""), true
	}
	// 4. Any other eviction or a vanished node is the worker, not the build.
	if is(pod.Reason, "Evicted") || pod.Phase == "Unknown" || is(pod.Reason, "NodeLost") {
		return failed(LostWorker, pod.Message.Or("")), true
	}
	for _, e := range exits {
		if e.ExitCode == 0 {
			continue
		}
		// 5. The fetch init container failed.
		if e.Name == FetchContainer {
			return failed(SourceUnavailable, e.Message.Or("")), true
		}
		// 6. The build container failed.
		return failed(BuildError, e.Message.Or(fmt.Sprintf("%s exited with code %d", e.Name, e.ExitCode))), true
	}
	return nil, false
}

// resourceKill applies rules 1 and 2: the build was killed for memory or disk.
func resourceKill(pod PodObservation, exits []ContainerExit) (JobVerdict, bool) {
	// 1. The kernel's OOM killer is unambiguous.
	for _, e := range exits {
		if is(e.Reason, "OOMKilled") {
			return failed(OutOfMemory, ""), true
		}
	}
	// 2. An eviction for ephemeral storage, or a build that saw ENOSPC.
	disk := is(pod.Reason, "Evicted") && mentionsDisk(pod.Message)
	for _, e := range exits {
		disk = disk || (e.ExitCode != 0 && mentionsDisk(e.Message))
	}
	if disk {
		return failed(DiskFull, pod.Message.Or("")), true
	}
	return nil, false
}

// reportOf is the report of the build container that exited cleanly.
func reportOf(exits []ContainerExit) (BuildReport, bool) {
	for _, e := range exits {
		if e.Name != BuildContainer || e.ExitCode != 0 {
			continue
		}
		message, ok := e.Message.Get()
		if !ok {
			return BuildReport{}, false
		}
		var report BuildReport
		if err := json.Unmarshal([]byte(strings.TrimSpace(message)), &report); err != nil {
			return BuildReport{}, false
		}
		return report, true
	}
	return BuildReport{}, false
}

// progress is where a Job that neither ended nor failed stands.
func progress(pod PodObservation, exits []ContainerExit) JobVerdict {
	if secs, ok := pod.UnschedulableSecs.Get(); ok && secs >= UnschedulableLimitSecs {
		return failed(Unschedulable, pod.Message.Or(""))
	}
	fetching := false
	for _, c := range pod.Running {
		if c == BuildContainer || c == ScanContainer {
			return Building{}
		}
		fetching = fetching || c == FetchContainer
	}
	for _, e := range exits {
		fetching = fetching || e.Name == FetchContainer
	}
	if fetching {
		return Fetching{}
	}
	return Pending{}
}
